diff -rup linux-5.15.78-orig/drivers/gpio/Kconfig linux-5.15.78/drivers/gpio/Kconfig
--- linux-5.15.78-orig/drivers/gpio/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/gpio/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -112,6 +112,13 @@ config GPIO_MAX730X
 menu "Memory mapped GPIO drivers"
 	depends on HAS_IOMEM
 
+config GPIO_CNS3XXX
+        bool "CNS3XXX GPIO support"
+        depends on ARCH_CNS3XXX
+        help
+	  Say yes here to support CNS3XXX GPIO device.
+	  Note : Only partial support at the moment. No IRQs
+
 config GPIO_74XX_MMIO
 	tristate "GPIO driver for 74xx-ICs with MMIO access"
 	depends on OF_GPIO
diff -rup linux-5.15.78-orig/drivers/gpio/Makefile linux-5.15.78/drivers/gpio/Makefile
--- linux-5.15.78-orig/drivers/gpio/Makefile	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/gpio/Makefile	2022-11-16 10:13:20.946425155 +1100
@@ -46,6 +46,7 @@ obj-$(CONFIG_GPIO_BRCMSTB)		+= gpio-brcm
 obj-$(CONFIG_GPIO_BT8XX)		+= gpio-bt8xx.o
 obj-$(CONFIG_GPIO_CADENCE)		+= gpio-cadence.o
 obj-$(CONFIG_GPIO_CLPS711X)		+= gpio-clps711x.o
+obj-$(CONFIG_GPIO_CNS3XXX)		+= gpio-cns3xxx.o
 obj-$(CONFIG_GPIO_SNPS_CREG)		+= gpio-creg-snps.o
 obj-$(CONFIG_GPIO_CRYSTAL_COVE)		+= gpio-crystalcove.o
 obj-$(CONFIG_GPIO_CS5535)		+= gpio-cs5535.o
diff -rup linux-5.15.78-orig/drivers/irqchip/irq-gic.c linux-5.15.78/drivers/irqchip/irq-gic.c
--- linux-5.15.78-orig/drivers/irqchip/irq-gic.c	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/irqchip/irq-gic.c	2022-11-16 10:13:20.946425155 +1100
@@ -1131,6 +1131,7 @@ static const struct irq_domain_ops gic_i
 static const struct irq_domain_ops gic_irq_domain_ops = {
 	.map = gic_irq_domain_map,
 	.unmap = gic_irq_domain_unmap,
+	.alloc = gic_irq_domain_alloc,
 };
 
 static void gic_init_chip(struct gic_chip_data *gic, struct device *dev,
diff -rup linux-5.15.78-orig/drivers/leds/Kconfig linux-5.15.78/drivers/leds/Kconfig
--- linux-5.15.78-orig/drivers/leds/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/leds/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -122,6 +122,15 @@ config LEDS_BCM6358
 	  This option enables support for LEDs connected to the BCM6358
 	  LED HW controller accessed via MMIO registers.
 
+config LEDS_CNS3XXX
+	tristate "LED Support for CNS3XXX"
+	depends on LEDS_CLASS
+	depends on ARCH_CNS3XXX
+	default y
+	help
+	  This option enables support for on-chip LED drivers found on CNS3XXX.
+
+
 config LEDS_CPCAP
 	tristate "LED Support for Motorola CPCAP"
 	depends on LEDS_CLASS
diff -rup linux-5.15.78-orig/drivers/leds/Makefile linux-5.15.78/drivers/leds/Makefile
--- linux-5.15.78-orig/drivers/leds/Makefile	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/leds/Makefile	2022-11-16 10:13:20.946425155 +1100
@@ -21,6 +21,7 @@ obj-$(CONFIG_LEDS_BCM6358)		+= leds-bcm6
 obj-$(CONFIG_LEDS_BD2802)		+= leds-bd2802.o
 obj-$(CONFIG_LEDS_BLINKM)		+= leds-blinkm.o
 obj-$(CONFIG_LEDS_CLEVO_MAIL)		+= leds-clevo-mail.o
+obj-$(CONFIG_LEDS_CNS3XXX)		+= leds-cns3xxx.o
 obj-$(CONFIG_LEDS_COBALT_QUBE)		+= leds-cobalt-qube.o
 obj-$(CONFIG_LEDS_COBALT_RAQ)		+= leds-cobalt-raq.o
 obj-$(CONFIG_LEDS_CPCAP)		+= leds-cpcap.o
diff -rup linux-5.15.78-orig/drivers/mtd/devices/Kconfig linux-5.15.78/drivers/mtd/devices/Kconfig
--- linux-5.15.78-orig/drivers/mtd/devices/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/mtd/devices/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -95,6 +95,15 @@ config MTD_MCHP48L640
 	help
 	  This enables access to Microchip 48L640 EERAM chips, using SPI.
 
+config CIRRUS_DUAL_MTD_ENV
+        bool "Use two partition for env on CNS3XXX. Requires UBoot support"
+        depends on ARCH_CNS3XXX
+        default y
+        help
+	  This option allows the kernel to recognise and use the
+	  dual environment variable style partition system as used
+	  on systems such as a CNS3xxx based Seagate Central.
+
 config MTD_SPEAR_SMI
 	tristate "SPEAR MTD NOR Support through SMI controller"
 	depends on PLAT_SPEAR || COMPILE_TEST
diff -rup linux-5.15.78-orig/drivers/net/ethernet/cavium/Kconfig linux-5.15.78/drivers/net/ethernet/cavium/Kconfig
--- linux-5.15.78-orig/drivers/net/ethernet/cavium/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/net/ethernet/cavium/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -13,6 +13,33 @@ config NET_VENDOR_CAVIUM
 
 if NET_VENDOR_CAVIUM
 
+config CNS3XXX_ETH
+	tristate "Cavium CNS3xxx Ethernet support"
+	depends on ARCH_CNS3XXX
+	select PHYLIB
+	help
+	  Say Y here if you want to use built-in Ethernet ports
+	  on CNS3XXX processor.
+
+config CNS3XXX_ETHADDR_IN_FLASH
+        bool "CNS3XXX MAC address in flash"
+	depends on CNS3XXX_ETH && MTD
+	help
+     	  Say Y if you want to load MAC address from flash.
+
+choice
+        prompt "Load MAC address from ..."
+	depends on CNS3XXX_ETHADDR_IN_FLASH && !CIRRUS_DUAL_MTD_ENV
+	default CNS3XXX_MAC_IN_PARALLEL_FLASH
+
+config CNS3XXX_MAC_IN_PARALLEL_FLASH
+        bool "Parallel Flash"
+
+config CNS3XXX_MAC_IN_SPI_FLASH
+        bool "SPI Flash"
+
+endchoice
+
 config THUNDER_NIC_PF
 	tristate "Thunder Physical function driver"
 	depends on 64BIT && PCI
diff -rup linux-5.15.78-orig/drivers/net/ethernet/cavium/Makefile linux-5.15.78/drivers/net/ethernet/cavium/Makefile
--- linux-5.15.78-orig/drivers/net/ethernet/cavium/Makefile	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/net/ethernet/cavium/Makefile	2022-11-16 10:13:20.946425155 +1100
@@ -2,6 +2,7 @@
 #
 # Makefile for the Cavium ethernet device drivers.
 #
+obj-$(CONFIG_CNS3XXX_ETH) += cns3xxx_eth.o
 obj-$(CONFIG_NET_VENDOR_CAVIUM) += common/
 obj-$(CONFIG_NET_VENDOR_CAVIUM) += thunder/
 obj-$(CONFIG_NET_VENDOR_CAVIUM) += liquidio/
diff -rup linux-5.15.78-orig/drivers/rtc/Kconfig linux-5.15.78/drivers/rtc/Kconfig
--- linux-5.15.78-orig/drivers/rtc/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/rtc/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -1619,6 +1619,24 @@ config RTC_DRV_CADENCE
 	  To compile this driver as a module, choose M here: the
 	  module will be called rtc-cadence.
 
+menuconfig RTC_DRV_CNS3XXX
+	tristate "Cavium Networks CNS3XXX RTC"
+       	depends on ARCH_CNS3XXX
+        help
+          If you say Y here you will get access to RTC found on
+	  CNS3XXX SOCs.
+
+          To compile this driver as a module, choose M here: the
+          module will be called rtc-cns3xxx.
+ 
+if RTC_DRV_CNS3XXX
+        config RTC_DRV_CNS3XXX_LOG_IN_FLASH
+        bool "CNS3XXX RTC log in flash"
+        depends on MTD
+        help
+          Say Y if you want to save/load log via flash.
+endif # RTC_DRV_CNS3XXX
+
 config RTC_DRV_FTRTC010
 	tristate "Faraday Technology FTRTC010 RTC"
 	depends on HAS_IOMEM
diff -rup linux-5.15.78-orig/drivers/rtc/Makefile linux-5.15.78/drivers/rtc/Makefile
--- linux-5.15.78-orig/drivers/rtc/Makefile	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/rtc/Makefile	2022-11-16 10:13:20.946425155 +1100
@@ -39,6 +39,8 @@ obj-$(CONFIG_RTC_DRV_BQ4802)	+= rtc-bq48
 obj-$(CONFIG_RTC_DRV_BRCMSTB)	+= rtc-brcmstb-waketimer.o
 obj-$(CONFIG_RTC_DRV_CADENCE)	+= rtc-cadence.o
 obj-$(CONFIG_RTC_DRV_CMOS)	+= rtc-cmos.o
+obj-$(CONFIG_RTC_DRV_CNS3XXX)   += rtc-cns3xxx.o
+obj-$(CONFIG_RTC_DRV_CNS3XXX_LOG_IN_FLASH) += rtc-cns3xxx-nvedit.o
 obj-$(CONFIG_RTC_DRV_CPCAP)	+= rtc-cpcap.o
 obj-$(CONFIG_RTC_DRV_CROS_EC)	+= rtc-cros-ec.o
 obj-$(CONFIG_RTC_DRV_DA9052)	+= rtc-da9052.o
diff -rup linux-5.15.78-orig/drivers/spi/Kconfig linux-5.15.78/drivers/spi/Kconfig
--- linux-5.15.78-orig/drivers/spi/Kconfig	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/spi/Kconfig	2022-11-16 10:13:20.946425155 +1100
@@ -36,6 +36,20 @@ config SPI_DEBUG
 # MASTER side ... talking to discrete SPI slave chips including microcontrollers
 #
 
+config SPI_CNS3XXX
+       tristate "CNS3XXX SPI controller"
+       depends on ARCH_CNS3XXX && SPI_MASTER
+       select SPI_BITBANG
+       help
+         This enables using the CNS3XXX SPI controller in master
+	 mode.
+
+config SPI_CNS3XXX_DEBUG
+       bool "Debug support for CNS3XXX SPI drivers"
+       depends on SPI_CNS3XXX
+       help
+         Say "yes" to enable debug messaging
+
 config SPI_MASTER
 #	bool "SPI Master Support"
 	bool
diff -rup linux-5.15.78-orig/drivers/spi/Makefile linux-5.15.78/drivers/spi/Makefile
--- linux-5.15.78-orig/drivers/spi/Makefile	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/spi/Makefile	2022-11-16 10:13:20.946425155 +1100
@@ -36,6 +36,7 @@ obj-$(CONFIG_SPI_CADENCE)		+= spi-cadenc
 obj-$(CONFIG_SPI_CADENCE_QUADSPI)	+= spi-cadence-quadspi.o
 obj-$(CONFIG_SPI_CLPS711X)		+= spi-clps711x.o
 obj-$(CONFIG_SPI_COLDFIRE_QSPI)		+= spi-coldfire-qspi.o
+obj-$(CONFIG_SPI_CNS3XXX) 		+= spi-cns3xxx.o
 obj-$(CONFIG_SPI_DAVINCI)		+= spi-davinci.o
 obj-$(CONFIG_SPI_DLN2)			+= spi-dln2.o
 obj-$(CONFIG_SPI_DESIGNWARE)		+= spi-dw.o
diff -rup linux-5.15.78-orig/drivers/usb/host/ohci-hcd.c linux-5.15.78/drivers/usb/host/ohci-hcd.c
--- linux-5.15.78-orig/drivers/usb/host/ohci-hcd.c	2022-11-11 04:15:43.000000000 +1100
+++ linux-5.15.78/drivers/usb/host/ohci-hcd.c	2022-11-16 10:13:20.946425155 +1100
@@ -1271,6 +1271,11 @@ MODULE_LICENSE ("GPL");
 #define TMIO_OHCI_DRIVER	ohci_hcd_tmio_driver
 #endif
 
+#ifdef CONFIG_USB_CNS3XXX_OHCI
+#include "ohci-cns3xxx.c"
+#define CNS3XXX_OHCI_DRIVER         ohci_hcd_cns3xxx_driver
+#endif
+
 static int __init ohci_hcd_mod_init(void)
 {
 	int retval = 0;
@@ -1315,6 +1320,12 @@ static int __init ohci_hcd_mod_init(void
 		goto error_tmio;
 #endif
 
+#ifdef CNS3XXX_OHCI_DRIVER
+        retval = platform_driver_register(&CNS3XXX_OHCI_DRIVER);
+	if (retval < 0)
+		goto error_cns3xxx;
+#endif
+	
 	return retval;
 
 	/* Error path */
@@ -1322,6 +1333,12 @@ static int __init ohci_hcd_mod_init(void
 	platform_driver_unregister(&TMIO_OHCI_DRIVER);
  error_tmio:
 #endif
+
+#ifdef CNS3XXX_OHCI_DRIVER
+	platform_driver_unregister(&CNS3XXX_OHCI_DRIVER);
+ error_cns3xxx:
+#endif
+	
 #ifdef SM501_OHCI_DRIVER
 	platform_driver_unregister(&SM501_OHCI_DRIVER);
  error_sm501:
@@ -1351,6 +1368,9 @@ static void __exit ohci_hcd_mod_exit(voi
 #ifdef TMIO_OHCI_DRIVER
 	platform_driver_unregister(&TMIO_OHCI_DRIVER);
 #endif
+#ifdef CNS3XXX_OHCI_DRIVER
+	platform_driver_unregister(&CNS3XXX_OHCI_DRIVER);
+#endif
 #ifdef SM501_OHCI_DRIVER
 	platform_driver_unregister(&SM501_OHCI_DRIVER);
 #endif
