diff -rup linux-5.15.81-orig/include/linux/pgtable.h linux-5.15.81/include/linux/pgtable.h
--- linux-5.15.81-orig/include/linux/pgtable.h	2022-12-03 03:41:12.000000000 +1100
+++ linux-5.15.81/include/linux/pgtable.h	2022-12-03 13:08:44.001576966 +1100
@@ -58,11 +58,13 @@
  * because in such cases PTRS_PER_PxD equals 1.
  */
 
+#ifndef pte_index
 static inline unsigned long pte_index(unsigned long address)
 {
 	return (address >> PAGE_SHIFT) & (PTRS_PER_PTE - 1);
 }
 #define pte_index pte_index
+#endif
 
 #ifndef pmd_index
 static inline unsigned long pmd_index(unsigned long address)
@@ -93,6 +95,7 @@ static inline pte_t *pte_offset_kernel(p
 #define pte_offset_kernel pte_offset_kernel
 #endif
 
+#ifndef pte_offset_map
 #if defined(CONFIG_HIGHPTE)
 #define pte_offset_map(dir, address)				\
 	((pte_t *)kmap_atomic(pmd_page(*(dir))) +		\
@@ -101,6 +104,7 @@ static inline pte_t *pte_offset_kernel(p
 #else
 #define pte_offset_map(dir, address)	pte_offset_kernel((dir), (address))
 #define pte_unmap(pte) ((void)(pte))	/* NOP */
+#endif /* CONFIG_HIGH_PTE */
 #endif
 
 /* Find an entry in the second-level page table.. */
