 fs/ntfs3/fslog.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/fs/ntfs3/fslog.c b/fs/ntfs3/fslog.c
index 397ba6a956e7..f2629a9f2499 100644
--- a/fs/ntfs3/fslog.c
+++ b/fs/ntfs3/fslog.c
@@ -1181,8 +1181,9 @@ static int log_read_rst(struct ntfs_log *log, u32 l_size, bool first,
 			struct restart_info *info)
 {
 	u32 skip, vbo;
-	struct RESTART_HDR *r_page = ntfs_malloc(DefaultLogPageSize);
-
+        struct RESTART_HDR *r_page = ntfs_malloc(PAGE_SIZE > DefaultLogPageSize ?
+						 PAGE_SIZE :		
+						 DefaultLogPageSize);
 	if (!r_page)
 		return -ENOMEM;
 
