 fs/ntfs3/fslog.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/fs/ntfs3/fslog.c b/fs/ntfs3/fslog.c
index 06492f088..e16bf639f 100644
--- a/fs/ntfs3/fslog.c
+++ b/fs/ntfs3/fslog.c
@@ -1180,8 +1180,9 @@ static int log_read_rst(struct ntfs_log *log, u32 l_size, bool first,
 			struct restart_info *info)
 {
 	u32 skip, vbo;
-	struct RESTART_HDR *r_page = kmalloc(DefaultLogPageSize, GFP_NOFS);
-
+        struct RESTART_HDR *r_page = kmalloc(PAGE_SIZE > DefaultLogPageSize ?
+						 PAGE_SIZE : DefaultLogPageSize,		
+						 GFP_NOFS);
 	if (!r_page)
 		return -ENOMEM;
 
